src/Microsoft.ML.Featurizers/TimeSeriesImputer.cs

```diff
@@ -0,0 +1,637 @@
+            new TimeSeriesImputerEstimator(CatalogUtils.GetEnvironment(catalog), timeSeriesColumn, grainColumns, filterColumns, filterMode, imputeMode, suppressTypeErrors);
+    }
+
+    /// <summary>
+    /// Imputes missing rows and column data per grain, based on the dates in the date column. Operates on a filtered list of columns in the IDataView.
+    /// If a column is not imputed but rows are added then it will be filled with the default value for that data type. Currently only numeric columns are supported for imputation.
+    /// A new column is added to the schema after this operation is run. The column is called "IsRowImputed" and is a boolean value representing if the row was created as a result
+    /// of this transformer or not.
+    /// </summary>
+    /// <remarks>
+    /// <format type="text/markdown"><![CDATA[
+    ///
+    /// ###  Estimator Characteristics
+    /// |  |  |
+    /// | -- | -- |
+    /// | Does this estimator need to look at the data to train its parameters? | Yes |
+    /// | Input column data type | All Types |
+    /// | Output column data type | All Types |
+    ///
+    /// The <xref:Microsoft.ML.Transforms.TimeSeriesImputerEstimator> is not a trivial estimator and needs training.
+    ///
+    ///
+    /// ]]>
+    /// </format>
+    /// </remarks>
+    /// <seealso cref="TimeSeriesImputerExtensionClass.TimeSeriesImputer(TransformsCatalog, string, string[], ImputationStrategy)"/>
+    /// <seealso cref="TimeSeriesImputerExtensionClass.TimeSeriesImputer(TransformsCatalog, string, string[], string[], FilterMode, ImputationStrategy, bool)"/>
+    public sealed class TimeSeriesImputerEstimator : IEstimator<TimeSeriesImputerTransformer>
+    {
+        private Options _options;
+        internal const string IsRowImputedColumnName = "IsRowImputed";
+
+        private readonly IHost _host;
+        private static readonly List<Type> _currentSupportedTypes = new List<Type> { typeof(sbyte), typeof(byte), typeof(short), typeof(ushort), typeof(int), typeof(uint),
+            typeof(long), typeof(ulong), typeof(float), typeof(double), typeof(string), typeof(ReadOnlyMemory<char>)};
+
+        #region Options
+        internal sealed class Options : TransformInputBase
+        {
+            [Argument(ArgumentType.Required, HelpText = "Column representing the time", Name = "TimeSeriesColumn", ShortName = "time", SortOrder = 1)]
+            public string TimeSeriesColumn;
+
+            [Argument((ArgumentType.MultipleUnique | ArgumentType.Required), HelpText = "List of grain columns", Name = "GrainColumns", ShortName = "grains", SortOrder = 2)]
+            public string[] GrainColumns;
+
+            // This transformer adds columns
+            [Argument(ArgumentType.MultipleUnique, HelpText = "Columns to filter", Name = "FilterColumns", ShortName = "filters", SortOrder = 2)]
+            public string[] FilterColumns;
```

Comment:
What's the use case for filtering? Are more columns produced besides `IsImputed`? Does the filtering operate only on the newly produced columns, or can I remove any pre-existing column?

Reply:
If a column is filtered, whenever a row is imputed the `default` value for that row is used to fill in the value. Using the `FilterMode`, you can either use this list as an include or exclude list.

Reply:
> If a column is filtered, whenever a row is imputed the default value for that row is used to fill in the value.

I'm not sure I'm parsing this correctly, and I think I'm missing something obvious. If a column is filtered, why do we fill in a value and then drop the column? Are there other columns produced besides `IsImputed`? What other columns would a user use the filtering to keep/remove besides `IsImpted`?

Reply:
We aren't dropping any columns in this process. If a column is excluded from being imputed, the default value is provided when a new row is generated. The biggest use case for this is dealing with columns that aren't supported for imputation. You can exclude that column from being imputed, but still have the default value filled in automatically.

Reply:
We may want to rename it from `FilterColumns` to perhaps `ColumnsToImpute` or `ColumnsToFillDefault`.  

The naming suggestion is to disambiguate from the existing concept in `ML.NET` of _filtering_, which removes rows of data from the IDataView ([info](https://docs.microsoft.com/en-us/dotnet/machine-learning/how-to-guides/prepare-data-ml-net#filter-data)).

Would another way of handling the non-supported column types be to automatically fill w/ default when needed? Seems we could auto-handle this use case. Assuming there aren't other use cases, this would remove the need of the `FilterColumns` field.

Reply:
I think renaming is a good idea. Its currently setup to where you can either list the columns to fill with default or columns to impute based on the `FilterMode` paramenter. So the name needs to somehow show that it can do both based on that parameter.

We could automatically fill in default values when needed. I didn't do that as this could potentially mask errors if you wanted to convert something prior and forgot. If you don't think thats a big deal though and think it would be better then I am not opposed to switching to that.

Reply:
Auto-filling w/ default for unsupported types is likely good from the data science side. Then we note in the docs "TimeSeriesImputer supports these data types (...) and will fill with the [default value](https://github.com/dotnet/machinelearning/blob/0a5514b8d1a8a597976aecec6f463e4471b73d6d/docs/code/IDataViewTypeSystem.md#default-value) for columns of unsupported data types".

@EricWrightAtWork -- since you know more about the needs of time-series forecasting, any thoughts? Is auto-filling unsupported datatypes with their default values reasonable?